### PR TITLE
Latex drawer tested using notebook in binder

### LINKED ISSUE #6371: Testing Latex drawer using binder
### What is the expected enhancement?

This is a follow on of PR #4544 . #4544 adds testing of MPL drawer using notebook in binder. As suggested in that PR, a similar form of testing is needed in order to improve the latex drawer testing.

### PR DESCRIPTION
**Pull request by TharrmashasthaPV**
### Summary

Fixes #6371 .

### Details and comments

Following #4544 , this PR enable testing Latex drawer using notebook in Binder. The flow of the testing follows exactly from #4544 .  The binder of this PR can be accessed at [https://mybinder.org/v2/gh/TharrmashasthaPV/qiskit-terra/binder_test_branch?filepath=test%2Fipynb%2Flatex_tester.ipynb](https://mybinder.org/v2/gh/TharrmashasthaPV/qiskit-terra/binder_test_branch?filepath=test%2Fipynb%2Fsnapshot_tester.ipynb). The release notes have been added.

### Files added/modified

- Renamed ```test/ipynb/mpl_tester.ipynb``` to ```test/iplynb/snapshot_tester.ipynb``` - The notebook that will run in Binder that includes tests for both mpl and latex drawers.
- Added ```apt.txt``` - This file installs the required latex dependencies. 
- Modified ```postBuild``` - This file now also collect and installs the latex qcircuit.sty package.
- Added ```test/ipynb/latex/references``` - This folder contains all the reference images.
- Added ```test/ipynb/latex/test_circuit_latex_drawer.py``` - The file that contains all the latex tests.

### CODE REVIEW

**Comment by javabster** [comment]
Hey @TharrmashasthaPV thanks for all the work on this! 

If you take a look at #6672 you can see how the `results.py` was adjusted to deal with both circuit and graph type snapshots. I think you can adjust the `results.py` file in a similar way here so that we don't have to duplicate the whole file just for the latex tests.

On another note, since we are expanding the snapshot testing I wonder if we should keep the `mpl_tester.ipynb` and `latex_tester.ipynb` separate or combine them into a single `snapshot_tester.ipynb`. My gut feel is put them all in one notebook, but open to further discussion 🤔

**Comment by TharrmashasthaPV** [comment]
> Hey @TharrmashasthaPV thanks for all the work on this!
> 
> If you take a look at #6672 you can see how the `results.py` was adjusted to deal with both circuit and graph type snapshots. I think you can adjust the `results.py` file in a similar way here so that we don't have to duplicate the whole file just for the latex tests.
> 
> On another note, since we are expanding the snapshot testing I wonder if we should keep the `mpl_tester.ipynb` and `latex_tester.ipynb` separate or combine them into a single `snapshot_tester.ipynb`. My gut feel is put them all in one notebook, but open to further discussion 🤔

Hi @javabster . That makes sense. But a small thing is that if we also put the latex testing in `results.py` it would force us to put the mpl and the latex tests together in a single notebook since calling `results.py` in the notebook would essentially run both mpl and latex tests together. If we want to have separate notebooks, alternatively, we could define different functions in `results.py` for all three different test and then use them in the notebooks appropriately. I don't have any strong preferences here. So any idea on which would be better?

**Comment by javabster** [comment]
Hey Tharma if I understand correctly you need to run the actual test file (e.g. `%run "mpl/circuit/test_circuit_matplotlib_drawer.py"`) in the notebook first to populate the `test_results.json`, and then that json is what is read by `results.py`. In which case if we wanted to keep 2 separate notebooks we could still use one  `results.py` and only call the corresponding test file (so `%run "mpl/circuit/test_circuit_matplotlib_drawer.py"` in the `mpl_tester.ipynb` and `%run 'latex/test_circuit_latex_drawer.py'` in the `latex_tester.ipynb`)

**Comment by javabster** [comment]
As for the 1 notebook vs 2 notebook decision I am on the fence 🤔 

On the one hand, keeping it all in one notebook means that you only have to open one binder link instead of 2 when you want to run the snapshot tests (and the binders take some time to load so fewer binder links the better maybe?)

On the other hand, if you're making changes to only the latex drawer maybe it's not necessary to run the mpl snapshots as well (and vice versa), in which case maybe it's better to keep them in separate notebooks.

@TharrmashasthaPV and @enavarro51 as you both have done a lot of work in the respective drawers, which do you feel would work best in your workflows?

**Comment by enavarro51** [comment]
@TharrmashasthaPV While looking at this PR in order to answer @javabster's query above, I took a look at some of the images, and when I download and view these in my viewer, they seem to be fairly low-res and are cropped around the edges. Is this something that can be fixed?
image

**Comment by enavarro51** [comment]
@javabster @TharrmashasthaPV In regards to whether to combine, I'm generally in favor of a single image tester, for mpl, latex, and graph (and maybe someday pulse as well). When working on only one drawer, the fact that the other images come up should not be an issue and should not increase load time significantly. And when working on multiple drawers, it would be helpful to see all at once.

In fact, though maybe not for this PR, I think that the tests for latex and mpl should be all in one file. They only differ by the `circuit_drawer `call in most cases. There are some, like `iqx_colors `that are mpl only, but almost all others are in common. So doing two `circuit_drawer `calls per test would eliminate having to add tests in 2 places. So in the best of all possible worlds, I see this.
```
ipynb
    image_tester.ipynb
    results.py
    circuit
        test_circuit_drawer.py
        references
    graph
        test_graph_drawer.py
        references
```

**Comment by TharrmashasthaPV** [comment]
> @TharrmashasthaPV While looking at this PR in order to answer @javabster's query above, I took a look at some of the images, and when I download and view these in my viewer, they seem to be fairly low-res and are cropped around the edges. Is this something that can be fixed?
> image

@enavarro51 I think the low resolution of the images are because the scale at which it is printed is small. The scale is exactly as would be drawn in a notebook. If we need larger resolved images, then I think we might have to change the value of `scale` in the drawer for the tests which I am not sure if we would want to do. I am open to any suggestions on that front. As for the cropping, it seems like an issue in `utils._trim()`. The method crops the image very close to the boundaries. I will fix that in my next commit.

@javabster I too agree with @enavarro51 in putting them into a single notebook. You were right in saying that we might not have to run the mpl snapshot tests every time we are testing the latex tests. I guess we can tackle that if we have separate methods for mpl_circuit, mpl_graph and latex in `results.py`, so that we can call only the required method.

**Comment by enavarro51** [comment]
@TharrmashasthaPV Looks like you got the cropping issue taken care of. It would be nice if the latex images were higher res, but I don't know how to do that through jupyter either. I know the mpl images come out much better res. Ultimately I don't suppose it's a problem if you're comparing one low-res image with another. They should still match if produced by the same code.

**Comment by javabster** [comment]
Hey @TharrmashasthaPV I just pushed a small change to make the results function a generic one that takes an argument instead of having 3 separate functions to do similar things and also tidied up the notebook a bit.

File: test/ipynb/latex/test_circuit_latex_drawer.py [hunk 1]
```diff
@@ -15,99 +15,120 @@
...
         circuit = QuantumCircuit(5)
         for qubit in range(5):
             circuit.h(qubit)
 
-        circuit_drawer(circuit, filename=filename, output="latex_source")
-
-        self.assertEqualToReference(filename)
+        self.circuit_drawer(circuit, filename="test_normal_circuit.png")
 
     def test_4597(self):
```
Comment (javabster) [inline test/ipynb/latex/test_circuit_latex_drawer.py:103]: This is not a very nice test name, we should have caught this earlier sorry! Would you ming renaming it as part of this PR?

**Review by javabster — COMMENTED** [review]
This is getting there thanks @TharrmashasthaPV! 

As you've changed the name of the binder test to `snapshot_tester` can you update the contributing guidelines to reflect the new name that people will have to put in the links they use

**Comment by TharrmashasthaPV** [comment]
> Hey @TharrmashasthaPV I just pushed a small change to make the results function a generic one that takes an argument instead of having 3 separate functions to do similar things and also tidied up the notebook a bit.

That is great! Thanks.

**Comment by TharrmashasthaPV** [comment]
> This is getting there thanks @TharrmashasthaPV!
> 
> As you've changed the name of the binder test to `snapshot_tester` can you update the contributing guidelines to reflect the new name that people will have to put in the links they use

Yeah, sure. I will make sure to update that. :)

File: test/ipynb/latex/test_circuit_latex_drawer.py (hunk 1 above)
Comment (TharrmashasthaPV) [reply test/ipynb/latex/test_circuit_latex_drawer.py]: I also did a similar change in one of the mpl circuit tests. I hope that's fine.

**Review by javabster — DISMISSED** [review]
This LGTM! 🚀 

@1ucian0 thoughts?

File: azure-pipelines.yml [hunk 2]
```diff
@@ -249,7 +249,7 @@ stages:
             image_tests/bin/pip install -U -r requirements.txt -c constraints.txt
             image_tests/bin/pip install -U -c constraints.txt -e ".[visualization]"
             image_tests/bin/python setup.py build_ext --inplace
-            sudo apt install -y graphviz pandoc
+            sudo apt install -y graphviz pandoc $(cat apt.txt)
```
Comment (1ucian0) [inline azure-pipelines.yml:252]: why these is needed? Do this checks run in CI?

File: environment.yml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+
```
Comment (1ucian0) [inline environment.yml:1]: is this file needed by binder? How is it different than `postBuild`?

File: releasenotes/notes/binder-for-latex-tests-2acdaf7d4a2aee4b.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+---
```
Comment (1ucian0) [inline releasenotes/notes/binder-for-latex-tests-2acdaf7d4a2aee4b.yaml:1]: I dont think there is a need of release note, since this change does not affect the users in any way.

File: test/ipynb/snapshot_tester.ipynb [hunk 5]
```diff
@@ -3,35 +3,61 @@
   {
    "cell_type": "code",
    "execution_count": null,
-   "id": "fc250437",
```
Comment (1ucian0) [inline test/ipynb/snapshot_tester.ipynb:6]: It is possible to "clean" the notebook with `nb-clean`. In this way, we can keep the minimum file in the repo.

**Review by 1ucian0 — COMMENTED** [review]
Awesome work.

For a following PR, would be possible to have a `snapshot_tester.ipynb` that calls  `mpl_tester.ipynb` and `latex_tester.ipynb`? I think the contributor only working a particular drawer can ran its own tester. For example, somebody working with MPL now needs to install all the dependencies for latex, when it is not needed. With https://github.com/Qiskit/qiskit-terra/pull/6427 , CI could run `snapshot_tester.ipynb`. What do you think?

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -327,12 +327,12 @@ we used in our CI systems more closely.
...
-you will need to check that your changes don't break any snapshot tests, and add
+If you are working on code that makes changes to any matplotlib or latex visualisations 
+you will need to check that your changes don't break any snapshot tests, and add 
 new tests where necessary. You can do this as follows:
 
 1. Make sure you have pushed your latest changes to your remote branch.
-2. Go to link: `https://mybinder.org/v2/gh/<github_user>/<repo>/<branch>?urlpath=apps/test/ipynb/mpl_tester.ipynb`. For example, if your GitHub username is `username`, your forked repo has the same name the original, and your branch is `my_awesome_new_feature`, you should visit https://mybinder.org/v2/gh/username/qiskit-terra/my_awesome_new_feature?urlpath=apps/test/ipynb/mpl_tester.ipynb.
+2. Go to link: `https://mybinder.org/v2/gh/<github_user>/<repo>/<branch>?urlpath=apps/test/ipynb/snapshot_tester.ipynb`. For example, if your GitHub username is `username`, your forked repo has the same name the original, and your branch is `my_awesome_new_feature`, you should visit https://mybinder.org/v2/gh/username/qiskit-terra/my_awesome_new_feature?urlpath=apps/test/ipynb/snapshot_tester.ipynb
```
Comment (1ucian0) [inline CONTRIBUTING.md:335]: I was not able to run https://mybinder.org/v2/gh/TharrmashasthaPV/qiskit-terra/binder_test_branch?filepath=apps/test/ipynb/snapshot_tester.ipynb . Maybe binder is in a bad day? Maybe I'm doing something wrong?

Comment (TharrmashasthaPV) [reply CONTRIBUTING.md:335]: There seems to be a small error in the link you tried. 😃 We use either `...filepath=test/ipynb/snapshot_tester.py` or `...urlpath=apps/test/ipynb/snapshot_tester.py` in the link. One correct link is [ https://mybinder.org/v2/gh/TharrmashasthaPV/qiskit-terra/binder_test_branch?urlpath=apps/test/ipynb/snapshot_tester.ipynb](https://mybinder.org/v2/gh/TharrmashasthaPV/qiskit-terra/binder_test_branch?urlpath=apps/test/ipynb/snapshot_tester.ipynb). This link seems to be working for me.

File: test/ipynb/snapshot_tester.ipynb (hunk 5 above)
Comment (TharrmashasthaPV) [reply test/ipynb/snapshot_tester.ipynb:6]: Yeah, sure. I'll do that 😃

File: releasenotes/notes/binder-for-latex-tests-2acdaf7d4a2aee4b.yaml (hunk 4 above)
Comment (TharrmashasthaPV) [reply releasenotes/notes/binder-for-latex-tests-2acdaf7d4a2aee4b.yaml:1]: Cool, I'll remove the reno.

File: azure-pipelines.yml (hunk 2 above)
Comment (TharrmashasthaPV) [reply azure-pipelines.yml:252]: I think images tests are run in CI at https://github.com/Qiskit/qiskit-terra/blob/0cdacec2ee3b26a96b1d7bbb0def6b7eb66ac5b4/azure-pipelines.yml#L255

File: environment.yml (hunk 3 above)
Comment (TharrmashasthaPV) [reply environment.yml:1]: This was required in one of the earlier versions. But I just figured that this was no longer required since the required packages are either installed using `postBuild` or `apt.txt` . Thanks for pointing this out! I'll remove this.

**Comment by TharrmashasthaPV** [comment]
> Awesome work.
> 
> For a following PR, would be possible to have a `py` version of `snapshot_tester.ipynb`? With #6427 , CI could run `snapshot_tester.py`. What do you think?

Thanks! I think currently we are just saving the output images at  
https://github.com/Qiskit/qiskit-terra/blob/0cdacec2ee3b26a96b1d7bbb0def6b7eb66ac5b4/azure-pipelines.yml#L255-L256
but not comparing them in CI. I think it certainly would be better to have comparison in CI and so I too feel we should have a `snapshot_tester.py`. I'll be sure to push it in a following PR.

File: CONTRIBUTING.md (hunk 6 above)
Comment (javabster) [reply CONTRIBUTING.md:335]: I also was struggling to run them a few days ago. I did manage to run them but they were very slow. Also wasn't sure if it was a binder issue, or maybe something to do with the changes here

Comment (TharrmashasthaPV) [reply CONTRIBUTING.md:335]: I just tested running branches with and without the change of this PR and I kind of found them both to take similar duration (around 3 mins) to open a server for the first time a branch is run after a commit. So, I think it is a binder issue.

**Review by javabster — DISMISSED** [review]
LGTM thanks Tharrma! 🚀

File: azure-pipelines.yml (hunk 2 above)
Comment (1ucian0) [reply azure-pipelines.yml:254]:
```suggestion
            sudo apt install -y graphviz pandoc texlive-pictures texlive-latex-recommended texlive-latex-extra texlive-fonts-recommended poppler-utils
```

File: postBuild [hunk 7]
```diff
@@ -11,6 +11,10 @@
 pip install matplotlib pylatexenc pillow appmode seaborn
 pip install .
 
+# Install QCircuit for LaTeX drawer
+mkdir -p $(kpsewhich -var-value=TEXMFHOME)/tex/latex/qcircuit/
+wget https://mirrors.ctan.org/graphics/qcircuit/qcircuit.sty -P $(kpsewhich -var-value=TEXMFHOME)/tex/latex/qcircuit/
```
Comment (1ucian0) [inline postBuild:16]:
```suggestion
tlmgr install qcircuit
```

File: test/ipynb/mpl/__init__.py [hunk 8]
```diff
@@ -10,7 +10,7 @@
 # copyright notice, and modified files need to carry a notice indicating
 # that they have been altered from the originals.
 
-"""Image comparison test for MPL graph drawer.
+"""Image comparison test for MPL drawer.
```
Comment (1ucian0) [inline test/ipynb/mpl/__init__.py:13]:
```suggestion
"""Image comparison test for MPL circuit drawer.
```

Comment (javabster) [reply test/ipynb/mpl/__init__.py:13]: I think just `MPL drawer` is correct, as it tests both the graphs and the circuits, then each subfolder has it's own `__init.py__` where it specifies if it's testing the graphs or circuits

Comment (1ucian0) [reply test/ipynb/mpl/__init__.py:13]: In that case... 

```suggestion
"""Image comparison test for MPL drawers.
```

**Comment by jakelishman** [comment]
In principle, despite its age, this PR still looks mostly appropriate to me, and not too far off merging.  I suspect that the conflicts are mostly a minor change in this PR conflicting with another minor styling change in the LaTeX output since this PR was opened.

@1ucian0 - you assigned yourself before.  Can you finish off driving this over the line?  If you write the small changes you think were required, I can do the final review.

**Comment by 1ucian0** [comment]
This can probably be closed as we are moving to CI visual testing in #9961
